File: src/08.dbpointer/11.83.go
package dbpointer

type ListNode struct {
	Val  int
	Next *ListNode
}

func IsSameList(l1 *ListNode, l2 *ListNode) bool {
	if l1 == nil && l2 == nil {
		return true
	}
	if l1 == nil || l2 == nil || l1.Val != l2.Val {
		return false
	}
	return IsSameList(l1.Next, l2.Next)
}

func CreateLintNode(nums []int) *ListNode {
	var head *ListNode
	var lastNode *ListNode

	for _, v := range nums {
		node := &ListNode{
			Val: v,
		}

		if head == nil {
			head = node
		} else {
			lastNode.Next = node
		}
		lastNode = node
	}

	return head
}

func DeleteDuplicates(head *ListNode) *ListNode {
	curNode := head
	for curNode != nil && curNode.Next != nil {
		if curNode.Val == curNode.Next.Val {
			curNode.Next = curNode.Next.Next
		} else {
			curNode = curNode.Next
		}
	}
	return head
}

File: src/08.dbpointer/11.83.ts
export class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

export function CreateListNode(nums: number[]): ListNode | null {
    let head: ListNode | null = null;
    let lastNode: ListNode;

    for (const v of nums) {
        const node = new ListNode(v);
        if (!head) {
            head = node;
        } else {
            lastNode.next = node;
        }
        lastNode = node;
    }

    return head;
}

export function IsSameListNode(
    l1: ListNode | null,
    l2: ListNode | null
): boolean {
    if (l1 === null && l2 === null) {
        return true;
    }

    if (l1 === null || l2 === null || l1.val !== l2.val) {
        return false;
    }

    return IsSameListNode(l1.next, l2.next);
}

export function deleteDuplicates(head: ListNode | null): ListNode | null {
    let curNode = head;

    while (curNode && curNode.next) {
        if (curNode.val === curNode.next.val) {
            curNode.next = curNode.next.next;
        } else {
            curNode = curNode.next;
        }
    }

    return head;
}
